Hermes/appscale/hermes/stats/producers/taskqueue_stats.py

```diff
@@ -0,0 +1,240 @@
+import collections
+
+import sys
+
+import time
+from tornado import gen, httpclient
+
+from appscale.hermes.stats.converter import include_list_name, Meta
+
+# The endpoint used for retrieving node stats.
+from appscale.hermes.stats.producers import proxy_stats
+
+STATS_ENDPOINT = '/service-stats'
+
+
+class BadTaskqueueStatsFormat(ValueError):
+  pass
+
+
+@include_list_name('taskqueue.cumulative')
+@attr.s(cmp=False, hash=False, slots=True, frozen=True)
+class CumulativeStatsSnapshot(object):
+  """ Cumulative counters reported for each TaskQueue instance. """
+  total = attr.ib()
+  failed = attr.ib()
+  pb_reqs = attr.ib()
+  rest_reqs = attr.ib()
+
+
+@include_list_name('taskqueue.recent')
+@attr.s(cmp=False, hash=False, slots=True, frozen=True)
+class RecentStatsSnapshot(object):
+  """ Recent stats reported for each TaskQueue instance. """
+  total = attr.ib()
+  failed = attr.ib()
+  avg_latency = attr.ib()
+  pb_reqs = attr.ib()
+  rest_reqs = attr.ib()
+  by_pb_method = attr.ib()
+  by_rest_method = attr.ib()
+  by_pb_status = attr.ib()
+  by_rest_status = attr.ib()
+
+
+@include_list_name('taskqueue.instance')
+@attr.s(cmp=False, hash=False, slots=True, frozen=True)
+class InstanceStatsSnapshot(object):
+  """ Stats reported for each TaskQueue instance. """
+  ip_port = attr.ib()
+  start_timestamp_ms = attr.ib()
+  current_requests = attr.ib()
+  cumulative = attr.ib(metadata={Meta.ENTITY: CumulativeStatsSnapshot})
+  recent = attr.ib(metadata={Meta.ENTITY: RecentStatsSnapshot})
+
+
+@include_list_name('taskqueue.failure')
+@attr.s(cmp=False, hash=False, slots=True, frozen=True)
+class FailureSnapshot(object):
+  """ Failure reported for a TaskQueue instance. """
+  ip_port = attr.ib()
+  error = attr.ib()
+
+
+@include_list_name('taskqueue')
+@attr.s(cmp=False, hash=False, slots=True, frozen=True)
+class TaskqueueServiceStatsSnapshot(object):
+  """ Stats reported for TaskQueue service. """
+  utc_timestamp = attr.ib()
+  current_requests = attr.ib()
+  cumulative = attr.ib(metadata={Meta.ENTITY: CumulativeStatsSnapshot})
+  recent = attr.ib(metadata={Meta.ENTITY: RecentStatsSnapshot})
+  instances = attr.ib(metadata={Meta.ENTITY_LIST: InstanceStatsSnapshot})
+  instances_count = attr.ib()
+  failures = attr.ib(metadata={Meta.ENTITY_LIST: FailureSnapshot})
+
+
+class TaskqueueStatsSource(object):
+
+  IGNORE_RECENT_OLDER_THAN = 5*60*1000  # 5 minutes
+  REQUEST_TIMEOUT = 10  # Wait up to 10 seconds
+
+  @gen.coroutine
+  def fetch_stats_from_instance(self, ip_port):
+    url = "http://{ip_port}{path}?last_milliseconds={max_age}".format(
+      ip_port=ip_port, path=STATS_ENDPOINT,
+      max_age=self.IGNORE_RECENT_OLDER_THAN
+    )
+    request = httpclient.HTTPRequest(
+      url=url, method='GET', request_timeout=self.REQUEST_TIMEOUT
+    )
+    async_client = httpclient.AsyncHTTPClient()
+
+    # Send Future object to coroutine and suspend till result is ready
+    response = yield async_client.fetch(request, raise_error=False)
```

Comment:
Should this handle a connection refused on the socket (in case the server is not even up)?

Reply:
Good question. I ran a test:
```
In [1]: from tornado import httpclient
In [2]: client = httpclient.HTTPClient()
In [3]: client.fetch(httpclient.HTTPRequest("http://localhost:12345"), raise_error=False)
Out[3]: HTTPResponse(_body=None,buffer=None,code=599,effective_url='http://localhost:12345',error=error(111, 'Connection refused'),headers={},reason='Unknown',request=<tornado.httpclient.HTTPRequest object at 0x7f8f4e247990>,request_time=0.0030510425567626953,time_info={})
```
So tornado httpclient handles socket.error and reports it as HTTP response `599 Connection refused`